### ISSUE TITLE
consult xref preview fails with "void-function xref-builder-location--eieio-childp"

### ISSUE THREAD

**Issue by bigodel** [minad/consult#473]
Trying to preview `consult-xref` results fails with `void-function xref-builder-location--eieio-childp`. Indeed this function isn't defined, and I couldn't find it anywhere. It appears, from [xref.el](https://git.savannah.gnu.org/gitweb/?p=emacs.git;a=blob;f=lisp/progmodes/xref.el#l65), that this has been deprecated. I thought it might be that some other package installation is also pulling a newer version of `xref`, but I also checked the code on `/usr/share/emacs/27.2/lisp/progmodes/xref.el.gz` and it also has no mention of such function. Don't know if this is `consult` related or not.

## Steps to reproduce

1. Start Emacs with `emacs -q`
2. Evaluate the following:

```lisp
(setq package-archives
      '(("melpa" . "https://melpa.org/packages/")
	("elpa" . "https://elpa.gnu.org/packages/")))

(package-initialize)
(package-refresh-contents)
(package-install 'consult)
(package-install 'vertico)

(vertico-mode)

(setq xref-show-xrefs-function #'consult-xref      
      xref-show-definitions-function #'consult-xref
      debug-on-error t)
```

3. Try to find references for a function with `M-? (xref-find-references)`, e.g. look for `format`, on the project prompt, select `... (choose a dir)` and select your Emacs configuration folder (I assume you have at least an instance of `format` in your config). You can also force the preview with `M-.`, the error occurs and shows the backtrace.

### Expected behaviour

Usages of `format` are shown and, upon navigating through them, or forcing the preview with `M-.`, the preview is shown in the buffer.

### Actual behaviour

Instead of seeing the preview, `[Error in post-command-hook (consult--preview-post-command): (void-functoin xref-builder-location--eieio-childp)]` shows up in the minibuffer and, when forcing it with `M-.`, an error is thrown with a backtrace, similar to this:

```
Debugger entered--Lisp error: (error "Command attempted to use minibuffer while in minib...")
 read-from-minibuffer("Find definitions of: " nil (keymap ... nil xref--read-identifier-history nil nil)
 completing-read-default("Find definitions of: " #f(compiled-function (string pred action) #<bytecode 0x156051bcb899>) nil nil nil xref--read-identifier-history nil nil)
 completing-read("Find definitions of: " #f(compiled-function (string pred action) #<bytecode 0x156051bcb899>) nil nil nil xref--read-identifier-history nil)
 xref--read-identifier("Find definitions of: ")
 byte-code("\300\301!C\207" [xref--read-identifier "Find definitions of: "] 2)
 call-interactively(xref-find-definitions nil nil)
 command-execute(xref-find-definitions)
 read-from-minibuffer("Go to xref: " nil (keymap ... nil consult-xref--history nil nil)
 completing-read-default("Go to xref: " #f(compiled-function (str pred action) #<bytecode 0x156051b61981>) nil t nil consult-xref--history nil nil)
 completing-read("Go to xref: " #f(compiled-function (str pred action) #<bytecode 0x156051b61981>) nil t nil consult-xref--history nil nil)
 #f(compiled-function () #<bytecode 0x156051b410dd>)()
 consult--with-preview-1(any #f(compiled-function (cand restore) #<bytecode 0x156051b20c99>) #f(compiled-function (input cand) #<bytecode 0x156051b3bcf9>) #f(compiled-function (&rest args2) #<bytecode 0x156051b3eda5>) #f(compiled-function () #<bytecode 0x156051b410dd>))
 consult--read-1(... :prompt "Go to xref: " :history consult-xref--history :require-match t :sort nil :category xref-location :group consult-xref--group :state #f(compiled-function (cand restore) #<bytecode 0x156051b20c99>) :lookup consult--lookup-candidate :prompt "Select: " :preview-key ...)
 apply(consult--read-1 (...) (:prompt "Go to xref: " :history consult-xref--history :require-match t :sort nil :category xref-location :group consult-xref--group :state #f(compiled-function (cand restore) #<bytecode 0x156051b20c99>) :lookup consult--lookup-candidate :prompt "Select: " :preview-key any ...))
 consult--read(... :prompt "Go to xref: " :history consult-xref--history :require-match t :sort nil :category xref-location :group consult-xref--group :state #f(compiled-function (cand restore) #<bytecode 0x156051b20c99>) :lookup consult--lookup-candidate)
 apply(consult--read (...) (:prompt "Go to xref: " :history consult-xref--history :require-match t :sort nil :category xref-location :group consult-xref--group :state #f(compiled-function (cand restore) #<bytecode 0x156051b20c99>) :lookup consult--lookup-candidate))
 consult-xref(#f(compiled-function () #<bytecode 0x156054a5e259>) ((window . #<window 3 on *scratch*>) (display-action) (auto-jump)))
 xref--show-xrefs(#f(compiled-function () #<bytecode 0x156054a5e259>) nil)
 xref--find-xrefs("format" references "format" nil)
 xref-find-references("format")
 funcall-interactively(xref-find-references "format")
 call-interactively(xref-find-references nil nil)
 command-execute(xref-find-references)
  ```

**Comment by minad** [comment]
I cannot reproduce this issue. My recommendation is to remove and reinstall consult and remove and reinstall xref. You may want to use the newest xref version from ELPA, it should work well. The stacktrace you showed is unrelated, it is an error due to recursive minibuffers. Try `(setq enable-recursive-minibuffers t)`. Furthermore the keybinding `M-.` is not defined in your above snippet.

**Comment by bigodel** [comment]
> My recommendation is to remove and reinstall Consult and remove and reinstall xref.

tried it and it didn't solve the issue.

> You may want to use the newest xref version from ELPA, it should work well.

i'm using Consult version 0.13 and xref, from ELPA, version 1.3.2.

> Note that the stacktrace you showed is unrelated, it is an error due to recursive minibuffers.

my bad, didn't really pay attention to it on the `emacs -q` environment, i have updated the description with the correct stacktrace.

**Comment by bigodel** [comment]
small side note: shouldn't the symbol `any`, in the code snippet on [this section](https://github.com/minad/consult#live-previews), be quoted? `(list ... :debounce 1 any)` throws a `void-variable` error

**Comment by minad** [comment]
I cannot reproduce the issue.

> small side note: shouldn't the symbol any, in the code snippet on this section, be quoted? (list ... :debounce 1 any) throws a void-variable error

Indeed.

**Comment by bigodel** [comment]
> I cannot reproduce the issue.

that's odd, it happens to me on a bare bones Emacs... i'm using Emacs 27.2 from `extra` on Arch Linux (the default package when you `pacman -S emacs`), is there any more information that might be helpful to someone else who might come to experience/investigate this issue?

**Comment by minad** [comment]
Try to dig a bit deeper. If you look into xref-1.3.2 there is no mentioning of eieio. So I wonder where xref-builder-location--eieio-childp comes from. The relevant code is here, in particular the typecase:

https://github.com/minad/consult/blob/37096d4bc4031a6a3d020ae225b6c675fa5ec0b7/consult-xref.el#L50-L72

Some systematic approaches:

1. Always work in clean environments. Delete .config/elpa.
2. Try emacs -Q (27.2) with consult from ELPA and xref from Emacs (do not install xref from ELPA, EIEIO version). Do you see the issue?
3. Try emacs -Q (28.1) with consult from ELPA and xref from Emacs (do not install xref from ELPA, non-EIEIO). Do you see the issue?
4. Try emacs -Q (27.2) with consult from ELPA and xref-1.3.2 from ELPA (non-EIEIO). Do you see the issue?
5. Try emacs -Q (28.1) with consult from ELPA and xref-1.3.2 from ELPA (non-EIEIO). Do you see the issue?

You may want to investigate the relevant generated code/bytecode. Where do you see eieio-childp?

**Comment by minad** [comment]
One more comment - your stack trace shows that `xref-buffer-location--eieio-childp` is executed on an xref struct. xref structs are the new post-EIEIO format introduced in xref-1.3. This hints at some miscompilation or presence of outdated code in your setup.

~~~ elisp
xref-buffer-location--eieio-childp(#s(xref-file-location :file "/home/foo/.config/emacs..." :line 143 :column 7))
 ~~~

**Comment by minad** [comment]
Also try to reorder the package installation order. xref should probably come first.
~~~ elisp
(package-install 'xref)
(package-install 'consult)
(package-install 'vertico)
~~~

**Comment by bigodel** [comment]
ok so it turns out that i was calling `(straight-use-package 'package)` manually, in order to get the most recent version of `project.el` without having to install Emacs 28 before the final release. that was pulling the most recent version of xref as well, though i'm not really sure why since the version of xref from Emacs that i had installed met its requirement. thank you for your suggestions, it led me to 1. solve my problem and 2. bite the bullet and install Emacs 28, so damn snappy!!!!

**Comment by bigodel** [comment]
although 2. has led me to a new error on `consult--invisible-open-temporarily`. it appears, from looking briefly at [the code](https://github.com/minad/consult/blob/37096d4bc4031a6a3d020ae225b6c675fa5ec0b7/consult.el#L1090) that Consult is trying to call `hs-isearch-show-temporary`, which is set as the overlay's `isearch-show-invisible-temporary`, with the wrong number of arguments, it expects an overlay and a "boolean". looking at other `isearch-show-invisible-temporary` functions, and the documentation on [Invisible Text](https://www.gnu.org/software/emacs/manual/html_node/elisp/Invisible-Text.html) (second to last paragraph), i believe the first argument should always be an overlay. i decided to comment here instead of opening a new issue, because this could a mistake on hideshow's part, but i'm not really sure. it is reproducible with `emacs -q` (again, in my machine).

```
Debugger entered--Lisp error: (wrong-number-of-arguments #<subr hs-isearch-show-temporary> 1)
  hs-isearch-show-temporary(nil)
  consult--invisible-open-temporarily()
```

**Comment by minad** [comment]
Please open a separate issue. This reminds me of https://github.com/raxod502/ctrlf/issues/100. I am not using hs-minor-mode and it does not seem to be widely used otherwise we would have already seen the issue :)

**Comment by bigodel** [comment]
> (EDIT: Fixed in [5d7c709](https://github.com/minad/consult/commit/5d7c709ebd61174c5af04610be1445393a877eab), no separate issue necessary).

thank you! i actually had solved it that way and just forgot to comment it here lol. thank you very much for everything, but mostly for being to attentive and responsive (besides the awesome packages and etc)
